tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-steps-selector.html

```diff
@@ -0,0 +1,194 @@
+      // asynchronously changing _runToStepIndex could create a rendering loop
+      // and jittery sliders.
+      setTimeout(() => {
+        this._updateSliders(this._runToStepIndex);
+      }, 0);
+    },
+    _computeColorForRun(run) {
+      return runsColorScale(run);
+    },
+    _computeTimeTextForRun(
+        runToAvailableTimeEntries, runToStepIndex, run, timeDisplayType) {
+      const stepIndex = runToStepIndex[run];
+      if (!_.isNumber(stepIndex)) {
+        // The step is not known yet.
+        return '';
+      }
+
+      const entries = runToAvailableTimeEntries[run];
+      if (!entries) {
+        // No PR curve data has been received from the server yet.
+        return '';
+      }
+
+      const value = entries[stepIndex][timeDisplayType];
+      if (timeDisplayType === 'step') {
+        return `step ${value}`;
+      } else if (timeDisplayType === 'relative') {
+        // Return the time using the units that are most apt.
+        if (value < 1) {
+          return `${(value * 1000).toFixed(2)} ms`;
+        }
+        return `${value.toFixed(2)} s`;
+      } else if (timeDisplayType === 'wall_time') {
+        return new Date(value * 1000).toString();
+      }
+      console.error(
+          `The display type of ${timeDisplayType} is not recognized.`);
+    },
+    _sliderValueChanged(event) {
+      const run = event.target.dataset.run;
+      const newRunToStepIndex = _.clone(this._runToStepIndex);
+      newRunToStepIndex[run] = event.target.immediateValue;
+      this.set('_runToStepIndex', newRunToStepIndex);
+    },
+    _computeMaxStepIndexForRun(runToAvailableTimeEntries, run) {
+      const entries = runToAvailableTimeEntries[run];
+      return entries && entries.length ? entries.length - 1 : 0;
+    },
+    _runToAvailableTimeEntriesChanged(runToAvailableTimeEntries) {
+      // The step for each run just refreshed. Change every slider to point to
+      // the last available step.
+      const newRunToStepIndex = {};
+      _.forOwn(runToAvailableTimeEntries, (entries, run) => {
+        if (entries.length) {
+          newRunToStepIndex[run] = entries.length - 1;
+        }
+      });
+      this.set('_runToStepIndex', newRunToStepIndex);
+
+      // Set the values of all the sliders. They are not 2-way bound, so we must
+      // actually set them.
+      this._updateSliders(newRunToStepIndex);
+    },
+    _updateSliders(runToStepIndex) {
+      const sliders = this.querySelectorAll("paper-slider");
+      if (sliders) {
+        for (let i = 0; i < sliders.length; i++) {
+          const run = sliders[i].dataset.run;
+          sliders[i].value = runToStepIndex[run];
+        }
+      }
+    },
+    _computeRunToStep(runToAvailableTimeEntries, runToStepIndex) {
+      const runToStep = {};
+      _.forOwn(runToStepIndex, (index, run) => {
+        const entries = runToAvailableTimeEntries[run];
+        if (!entries) {
+          return
```

Comment:
Missing semicolon.

I'm surprised that this doesn't raise a warning with the closure compiler?

Reply:
Done. Hmm, indeed, I see no Closure warnings. Maybe the transpilation from TS to uncompiled JS adds the semicolon back.

Reply:
But this isn't TypeScript?

Reply:
I just double-checked that no warning appears with no semicolon. Indeed, a warning should appear since suspiciousCode warnings are on by default: https://github.com/google/closure-compiler/wiki/Warnings

Maybe the TS is transpiled into JS before being compiled into a small JS binary, and Closure applies in the 2nd step?